packages/lib/hooks/useTypedQuery.ts

```diff
@@ -0,0 +1,59 @@
+import { useRouter } from "next/router";
+import { z } from "zod";
+
+export function useTypedQuery<T extends z.Schema>(schema: T) {
+  type InferedSchema = z.infer<typeof schema>;
+  type SchemaKeys = keyof InferedSchema;
+  const { query: unparsedQuery, ...router } = useRouter();
+  const parsedQuery = schema.parse(unparsedQuery) as InferedSchema;
+
+  // Set the query based on schema values
+  function setQuery<J extends SchemaKeys>(key: J, value: InferedSchema[J]) {
+    router.replace({ query: { ...parsedQuery, [key]: value } }, undefined, { shallow: true });
+  }
+
+  // Delete a key from the query
+  function removeByKey(key: SchemaKeys) {
```

Comment:
This is not safe since you allow to remove keys that are required, you likely wanna filter and only allow removal of optional keys